# Content

* 1. Import Libraries

* 2. Create Dataframe

* 3. Check for mixed-data type

* 4. Missing Values
    * 4.1 Find Missing Values
    * 4.2 How to deal with Missing values?
        * 4.2.1 Create subset with missing values
        * 4.2.2 Impute values
        * 4.2.3 Droppong all missing Values
        
* 5. Duplicates
    * 5.1 Finding Duplicats
    * 5.2 How to deal with Duplicates?
        * 5.2.1 Delete Duplicates
        
* 6. Exercise 4.5
    * 6.1 Check for Duplicates
    * 6.2 Check for data types
    
* 7. Exporting Data





# 1. Import Libraries

In [1]:
# import libraries
import pandas as pd
import numpy as np
import os

# 2. Create Dataframe

In [2]:
df_test=pd.DataFrame()

In [3]:
df_test['mixed']=['a','b',1,True]

# 3. Check for mixed-data type

In [4]:
## Command to check for mixed data type in out dataframe
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mixed


C:\Users\binth\AppData\Local\Temp\ipykernel_36544\134491531.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


 The Output says "mixed", which means this column has a mixed-data type

In [5]:
##create another dataframe
df_test2 =pd.DataFrame()
df_test2 = pd.DataFrame(
         {"a" : [1 ,2, 3],
          "b" : [7, 8, 'a'],
          "c" : [False, True, 'a']})

In [6]:
df_test2

a  b      c
0  1  7  False
1  2  8   True
2  3  a      a

In [7]:
for col in df_test2.columns.tolist():
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test2[weird]) > 0:
    print (col)

b
c


C:\Users\binth\AppData\Local\Temp\ipykernel_36544\2849338216.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Temp\ipykernel_36544\2849338216.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Temp\ipykernel_36544\2849338216.py:2: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test2[[col]].applymap(type) != df_test2[[col]].iloc[0].apply(type)).any(axis = 1)


# 4. Missing values

Importing the datasets we need

In [2]:
## import first our dataframe
#import original data of "product.csv" and prepared data of "orders_wrangled.csv"
ogpath=r'C:\Users\binth\Documents\CareerFoundry\Achievement 4 - Python\InstaCart\Instacart Business Analysis'

In [3]:
## import data
df_prods=pd.read_csv(os.path.join(ogpath,'02 Data','Original Data','products.csv'))

In [4]:
df_ords=pd.read_csv(os.path.join(ogpath,'02 Data','Prepared Data','orders_wrangled.csv'))

## 4.1. Finding Missing values

In [11]:
## from products.csv
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
## Create a subset where missing values are located
df_nan= df_prods[df_prods['product_name'].isnull()==True]

In [13]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# 4.2 How to deal with missing values?

## 4.2.1 Create a new variable only containing missing values information

## 4.2.2 Imputing Values

In [ ]:
## When values are numeric: impute with mean values and overwrite it

df['column with missings'].fillna(mean value, inplace=True)

In [ ]:
## When values are numeric: impute with median values and overwrite it

df['column with missings'].fillna(median value, inplace=True)

In [14]:
## Creating a new Dataframe without the missing values

df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

### 4.2.3 Drop all missing values

In [ ]:
## df_prods.dropna(inplace = True)
##df_prods.dropna(subset = [‘product_name’], inplace = True)

# 5. Duplicates

## 5.1 Finding duplicates

In [9]:
df_prods.duplicated().sum()

5

In [15]:
## creating a subset containing full duplicates only
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [21]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## 5.2 How to proceed with duplicates?

# 5.2.1 Delete Duplicates

In [16]:
# creating a dataframe where all duplicated is dropped
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

### Export the data

In [17]:
## Export as a new data
df_prods_clean_no_dups.to_csv(os.path.join(ogpath,'02 Data','Prepared Data','products_checked.csv'))

# 6. Exercise 4.5

In [18]:
## Run the df.describe() function on your df_ords dataframe. 
#anything about the data looks off or should be investigated further?
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week   time_of_day  days_since_previous_order  
count        3.421083e+06  3.421083e+06               3.214874e+06  
mean         2.776219e+00  1.345202e+01               1.111484e+01  
std          2.046829e+00  4.226088e+00               9.206737e+00  
min          0.000000e+00  0.000000e+00               0.000000e+00  
25%          1.000000e+00  1.000000e+01               4.000000e+00  
50%          3.000000e+00  1.300000e+01               7.000000e+00  
75%          5.000000e+00  1.600000e+01               1.500000e+01  
max          6.000000e+00  2.300000e+01               3.000000e+01

 Anything about the data looks off or should be investigated further?

Answer: Yes.

Somehow, order_id and user_id are still in int64 format. It should be in str. Order_number: 25% percentile shows 5, which is higher than max. Value (1) days_since_prio_prder: min. is 0 , max. is 3 BUT 50% percentile is 7. This seems off.


In [28]:
## Check for mixed-type data in your df_ords dataframe.
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

C:\Users\binth\AppData\Local\Temp\ipykernel_16316\847038604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Temp\ipykernel_16316\847038604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Temp\ipykernel_16316\847038604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Temp\ipykernel_16316\847038604.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
C:\Users\binth\AppData\Local\Tem

### Result: No mixed-data founded in all columns

In [19]:
## Changing the data type of different columns
df_ords['order_id']=df_ords['order_id'].astype('str')
df_ords['user_id']=df_ords['user_id'].astype('str')

In [20]:
df_ords.dtypes

Unnamed: 0                     int64
order_id                      object
user_id                       object
order_number                   int64
orders_day_of_week             int64
time_of_day                    int64
days_since_previous_order    float64
dtype: object

In [21]:
## Check for missing values
df_ords.isnull().sum()

Unnamed: 0                        0
order_id                          0
user_id                           0
order_number                      0
orders_day_of_week                0
time_of_day                       0
days_since_previous_order    206209
dtype: int64

**206209 missing values are located in the column of "days_since_prior_order". 
This might be caused by human error in the programme or human error? I would not know why the remaining orders are complete if this process is automated by a computer**.

**How should we proceed with the missing values???**

* Missing values of 206209 is approx. 6% of the entire observations (3421083) and could be exclude from the dataframe of df_ords. On the other hand, it is better to create a subset of NaNs and investigate it.

* It seems like the number of missing number of days_since_prior_order (206209) is the same number as the user_id (from 1 to 206209). This suggest that those are customers who have only purchased 1 item sofar and not more. The missing values can therefore be left out as blanks, in case the same user_id want to purchase another time.

## 6.1 Run a check for duplicate values in your df_ords data

In [22]:
df_dup_ords = df_ords[df_ords.duplicated()]

In [23]:
df_dup_ords

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, time_of_day, days_since_previous_order]
Index: []

### Result: No duplicates found in all columns

## 6.2 Check for all data types again before Exporting

* Creating a copy of the ords. and prods dataframes. This new dataframe will contain the changed datatype for the collumns, because overwriting with .astype() did not work.

In [29]:
## Creating a copy of ords.csv
df_ords2 = df_ords.copy()

In [30]:
# Creating a copy of df_prods_clean_no_dups
df_prods2 = df_prods_clean_no_dups.copy()

In [31]:
## Chaning data type in the new data frame
df_ords2['order_id']=df_ords2['order_id'].astype('str')
df_ords2['user_id']=df_ords2['user_id'].astype('str')

In [35]:
# changing data type in new prods dataframe
df_prods2['product_id']=df_prods2['product_id'].astype('str')
df_prods2['product_name']=df_prods2['product_name'].astype('str')
df_prods2['aisle_id']=df_prods2['aisle_id'].astype('str')
df_prods2['department_id']=df_prods2['department_id'].astype('str')

In [36]:
df_prods2.dtypes

product_id        object
product_name      object
aisle_id          object
department_id     object
prices           float64
dtype: object

In [38]:
df_ords2.dtypes

Unnamed: 0                     int64
order_id                      object
user_id                       object
order_number                   int64
orders_day_of_week             int64
time_of_day                    int64
days_since_previous_order    float64
dtype: object

In [39]:
df_prods2.dtypes

product_id        object
product_name      object
aisle_id          object
department_id     object
prices           float64
dtype: object

# 7. Exporting the data

In [45]:
## Export as a new data in the Prepared Data folder
df_ords2.to_csv(os.path.join(ogpath,'02 Data','Prepared Data','orders_checked.csv'))

TypeError: NDFrame.to_csv() got an unexpected keyword argument 'index_col'

In [46]:
df_prods2.to_csv(os.path.join(ogpath,'02 Data','Prepared Data','products_checked.csv'))

TypeError: NDFrame.to_csv() got an unexpected keyword argument 'index_col'